### DISCUSSION TITLE
`viewAny = false` policy overrides `view`

### DISCUSSION THREAD

**Discussion by bilogic** [filamentphp/filament#13157] [Help] [answered]
### Package

Table builder

### Package Version

3.2.86

### How can we help you?

I like my users to be able to update their own user info, but not allowed to view/update other users. Is there a way to achieve using policy alone?

1. When `viewAny = false`, view and update both won't work (view+update depends on viewAny)
2. If `viewAny` is meant to refer to index, why does it affect the view and update pages?
3. By the same logic, if we cannot view, why should we be allowed to update? (update does not depend on view)
4. In practice, this is `view = false`, `viewAny = true` and `update = true`, which leaves update enabled
5. I think it is better that these permissions (can/can't load) have no interdependence

**Comment by leandrocfe** [comment, marked as answer]
You could use Laravel Scopes to show records by user

```php
public function apply(Builder $builder, Model $model): void
{
    $user = auth()->user();
    $builder->whereUserId($user->id);
}
```

**Reply by bilogic** [reply]
My scopes are normally in the model's `boot()`. Which class should I apply this? thanks

**Reply by leandrocfe** [reply]
You can create scopes: `php artisan make:scope YourScope`
And use it in your model

```php
#[ScopedBy(YourScope::class)]
class YourModel extends Model
{
...
}
```

**Reply by bilogic** [reply]
oh, you meant a scope in a class of its own, ok got it thanks

Side note what I don't like about using scopes is that it scatters the ACL logic into multiple files, i.e. new developer is unaware of the existence of an ACL scope when reading the policy file.

**Reply by leandrocfe** [reply]
If you don't want to use scopes
https://filamentphp.com/docs/3.x/panels/resources/getting-started#customizing-the-resource-eloquent-query

**Reply by bilogic** [reply]
Haha, thanks, but it is the same situation, logic scattering, nevermind I will work with what there is

**Comment by kimselius** [comment]
This actually sucks and makes no sense at all..

There is no possibility to show a single record or make a user update a single record without allowing them to view a listing of all records?

Filters (globally) will work fine if you can scope them to some parameter in the Model or relationship, but if you use records where you don't have this type of scoping it's not possible.

1. Why does viewAny block access to everything? Makes no sense. 
2. Why is it not possible to override the access of the listing page except viewAny?
3. It's not possible to remove the listing resource? Everything depends on the index route to be registered?

**Comment by dharen008** [comment]
I'm facing the same issue. I have two panels: an admin panel and a client panel. Client registration happens in the admin panel, while in the client panel, they can only view their own profile page. I've already configured both panel providers, and there's nothing wrong with those files. The problem is that viewAny is overriding the view permission when it shouldn't. If viewAny overrides everything, what's the point of having separate permissions?

```
class CompanyPolicy
{
	/**
	 * Determine whether the user can view any models.
	 */
	public function viewAny( User|Company $user ): bool
	{
		if ( $user instanceof User ) {
			return $user->greaterThanAdmin();
		}

                // Logic for Company
		return false;
	}

	/**
	 * Determine whether the user can view the model.
	 */
	public function view( User|Company $user, Company $company ): bool
	{
		if ( $user instanceof User ) {
			return $user->greaterThanAdmin();
		}

                // Logic for Company
		return $user instanceof Company && $user->id === $company->id;
	}
}
```

**Comment by Patabugen** [comment]
I also ran into this with the use case that I want a user to be able to edit their own Organisation without being able to list all of them.

I get the "viewAny" permission being literally: "Are there any models they can view"

It would be great for Filament to be able to differentiate "viewAll" perhaps.

**Comment by Patabugen** [comment]
Here's my workaround for this issue, in case it's helpful for anyone.

I added an 'index' permission to the Policy, and expanded on the authentication methods added by `Filament\Resources\Resource\Concerns\HasAuthorization`.

So now you can use `canAccess() && ! canViewIndex()` to pick up someone who can view some models, but can't list them.

I've plucked this out of my working project, and hopefully haven't missed anything - happy to hear any feedback/questions.

I'm using this with Filament v4

```php
<?php

namespace App\Filament\Resources\Concerns;

use Illuminate\Auth\Access\Response;

/**
 * Filament doesn't have a way to differentiate between being able to display
 * the index listing of a resource and being able to view individual records.
 *
 * A user needs `viewAny` permissions to access the panel at all, but then there's
 * no way to let them modify their own Organisation (as an example Model) without letting them
 * access the list of all Organisations.
 *
 * To workaround this we've added a custom "index" permission, and this trait adds
 * relevant methods to the Resource.
 */
trait HasIndexAuth
{
    public static function canViewIndex(): bool
    {
        return static::getViewIndexAuthorizationResponse()->allowed();
    }

    public static function getViewIndexAuthorizationResponse(): Response
    {
        return static::getAuthorizationResponse('index');
    }
}
```

Add the `index` method to your policy:
```php
<?php

namespace App\Policies;

use App\Models\Organisation;

class OrganisationPolicy
{
    public function index(User $user): bool
    {
        return $user->can('organisations.index');
    }
}
```
# Example Usage
## Hide Navigation
Don't register the Navigation if a user doesn't have access to it (in my version, I actually have this in the trait)
```
<?php

namespace App\Filament\Resources\Organisations;

class OrganisationResource extends BaseResource
{

    // Only add the navigation is the user can view the index.
    public static function shouldRegisterNavigation(): bool
    {
        return static::canViewIndex();
    }
}
```
## Hide Breadcrumb on Edit Page
On the "Edit" page I override "hasResourceBreadcrumbs" so the breadcrumb is hidden if the user can't access the index:

```php
<?php

namespace App\Filament\Resources\Organisations\Pages;

class EditOrganisation extends EditRecord
{
    public function hasResourceBreadcrumbs(): bool
    {
        return self::getResource()::canViewIndex();
    }
}
```

## Disallow access to the List page

override `authorizeAccess` to check access.

```php
<?php

namespace App\Filament\Resources\Organisations\Pages;

use App\Filament\Resources\Organisations\OrganisationResource;
use Filament\Resources\Pages\ListRecords;

class ListOrganisations extends ListRecords
{
    protected function authorizeAccess(): void
    {
        abort_unless(
            return self::getResource()::canViewIndex();
            403
        );
    }
```
